### DISCUSSION TITLE
FastAPI multiple StreamRouters

### DISCUSSION THREAD

**Discussion by JoshJarabek7** [ag2ai/faststream#677] [Q&A] [answered]
I'm struggling to get around the circular dependencies issue for FastAPI.

app.main - I create the faststream.rabbit.fastapi.RabbitRouter and I include it as a router for the app.

app.user.api - I can't import the router here, because these routes won't get included in app even though "include_router" happens. I can use the broker() function as a dependency, but that doesn't matter, because I still need a router. If I import the router from app.user.api to app.main then it's a circular dependency issue, because I'm using broker() as a dependency injection.

The only way I'm getting this to work is if I keep all my API routes on app.main.

Otherwise, they don't get included in the app's routes, or I receive a circular dependency.

I believe this creates a broker each time the RabbitRouter is initialized, if not, then that makes things easy and I can just use it as a regular APIRouter for FastAPI. I was looking through the codebase, and it looks like there might be a high-level abstraction to aio-pika's Connection class to potentially draw from as a Singleton? If I'm able to just declare the Connection in its own module and then implement the connection through the brokers and routers that would help!

Any tips or advice?

Thanks! This project rocks!

**Comment by Lancetnik** [comment, marked as answer]
As I understood you correct, you are trying to create broker in the same with FastAPI application file (app.main) and declare some handler in the other file. But you can create the broker in the same with handlers file (app.user.api) and just import it in the app.main

**Reply by JoshJarabek7** [reply]
Doesn't that create an extra connection though for RabbitMQ with each broker created?

EDIT:

Oh sorry, I just implemented what you said. I had been trying to use the RabbitRouter in each api.py file for FastAPI. I just tried implementing the regular API router and initialize a RabbitBroker on each page instead, and the RabbitMQ connection count stays at 1 with multiple brokers connected.

Thanks!

**Reply by Lancetnik** [reply]
You can declare any number of  `StreamRouter` and include them in each other. If you finally includes just a one to `FastAPI` instance, you have the only one connection.

**Reply by JoshJarabek7** [reply]
Oh! Okay, that makes things easier. I'm going to try that. Thank you!

**Reply by JoshJarabek7** [reply]
@Lancetnik sorry for all the questions, but I'm still struggling to get 1 single connection, while getting the queues to work.

#app.main
```
from fastapi import FastAPI
import os
from faststream.rabbit.fastapi import RabbitRouter
from app.ai.api import router as ai_router
from app.user.api import router as user_router

RABBITMQ_URL = os.environ.get("RABBITMQ-URL")
router = RabbitRouter(url=RABBITMQ_URL)

app = FastAPI()

@app.on_event("startup")
async def startup_event():
    await router.broker.start()
    # await ai_router.broker.start()
    # await user_router.broker.start()
    print(ai_router.broker.started)

@app.on_event("shutdown")
async def shutdown_event():
    await router.broker.close()


router.include_router(ai_router)
router.include_router(user_router)
app.include_router(router)
```

#app.ai.api
```
from faststream.rabbit.fastapi import RabbitRouter
router = RabbitRouter()
@router.subscriber("ai.test.subscriber")
async def ai_test_subscriber(msg: dict):
    print(msg)

@router.get("/api/ai/test-ai")
async def test_ai():
    await router.broker.publish(
        message={"message": "test"}, queue="ai.test.subscriber"
    )
```

With this current setup, these are the startup logs:
```
INFO:     Started server process [4672]
INFO:     Waiting for application startup.
2023-09-20 15:47:47,932 INFO     - default | ai.test.subscriber       |            - `AiTestSubscriberAi.Test.Subscriber` waiting for messages
2023-09-20 15:47:48,024 INFO     - default | user.notification.create |            - `CreateUserNotificationUser.Notification.Create` waiting for messages
2023-09-20 15:47:48,117 INFO     - default | user.notification.read   |            - `ReadNotificationUser.Notification.Read` waiting for messages
2023-09-20 15:47:48,213 INFO     - default | user.create              |            - `CreateUserHandlerUser.Create` waiting for messages
2023-09-20 15:47:48,305 INFO     - default | user.deactivate          |            - `DeactivateUserHandlerUser.Deactivate` waiting for messages
False
INFO:     Application startup complete.
```

When hitting the test endpoint with Postman, I receive this:
```
  File "/Users/josh/OMEGA/app/ai/api.py", line 44, in test_ai
    await router.broker.publish(message={"message": "test"}, queue="ai.test.subscriber")
  File "/Users/josh/Library/Caches/pypoetry/virtualenvs/omega-T04Oa7NS-py3.11/lib/python3.11/site-packages/faststream/rabbit/broker.py", line 394, in publish
    assert self._producer, "RabbitBroker channel is not started yet"  # nosec B101
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: RabbitBroker channel is not started yet
```

However, my RabbitMQ dashboard is showing 1 connection, 1 channel, and 6 queues, which is good because that's what I'm looking for, and the queues show up in the startup logs, but it's throwing the error stating that the RabbitBroker channel is not started yet.

Now, if I uncomment the comments on main.py within the startup_event() function:

```
@app.on_event("startup")
async def startup_event():
    # await router.broker.start()
    await ai_router.broker.start()
    await user_router.broker.start()
    print(ai_router.broker.started)
```

This works and the message goes through, however, it creates a RabbitMQ connection for each router. How do I go about limiting the connections to 1 per application, while being able to have each of the RabbitRouters started?

Sorry, I'm a newbie to RabbitMQ.

**Reply by Lancetnik** [reply]
You should setup FastAPI lifespan with the  router one.
app = FastAPI(lifespan=router.lifespan_context)

**Reply by Lancetnik** [reply]
Use the core router (that includes others) to it

**Reply by JoshJarabek7** [reply]
The FastAPI(lifespan=router.lifespan_context) also did not work for me. I ended up creating a Singleton RabbitRouter that I just import into each of the api.py files and that ended up working, so that they're all just technically using the same router.

Which one is the core router? Is that the non-fastapi RabbitRouter?

**Reply by Lancetnik** [reply]
@JoshJarabek7 this is a working code example:

```python
from faststream.rabbit.fastapi import RabbitRouter
from fastapi import FastAPI

# some router in `api/router1`
sub_router_1 = RabbitRouter()

@sub_router_1.subscriber("sub1")
async def handler_sub1(msg):
    print(f"sub1: {msg=}")


# some router in `api/router2`
sub_router_2 = RabbitRouter()

@sub_router_2.subscriber("sub2")
async def handler_sub2(msg):
    print(f"sub2: {msg=}")


# some router in `api/router3`
core_router = RabbitRouter()

@core_router.subscriber("core")
async def handler_sub2(msg):
    print(f"core: {msg=}")


# main.py where all routers are imported
core_router.include_router(sub_router_1)
core_router.include_router(sub_router_2)

app = FastAPI(lifespan=core_router.lifespan_context)
app.include_router(core_router)


@core_router.after_startup
async def handle(app: FastAPI):
    await core_router.broker.publish("Hi!", "sub1")
    await core_router.broker.publish("Hi!", "sub2")
    await core_router.broker.publish("Hi!", "core")
```

**Reply by dimbler** [reply]
Dear @Lancetnik
Please clarify is it possible to send data between different routers? From router_source to router_dest in my case.
Today separate routers works good, but togaether got unexpected behaviour `kafka.errors.TopicAuthorizationFailedError: [Error 29] TopicAuthorizationFailedError: test-topic`
But connection established and verified, probably group_id missing during connection?
here is my code:

```python
logger.info(f"Source to {settings['kafka_servers_source']} as {settings['kafka_source.username']}")
kafka_security_source = SASLScram512(username=settings["kafka_source.username"], password=settings["kafka_source.password"], use_ssl=True, ssl_context=ssl_context_source)
router_source = KafkaRouter(bootstrap_servers=settings['kafka_servers_source'], security=kafka_security_source)

logger.info(f"Dest to {settings['kafka_servers_dest']} as {settings['kafka_dest.username']}")
kafka_security_dest = SASLScram512(username=settings["kafka_dest.username"], password=settings["kafka_dest.password"], use_ssl=True, ssl_context=ssl_context_dest)
router_dest = KafkaRouter(bootstrap_servers=settings['kafka_servers_dest'], security=kafka_security_dest)

async def handle(log_messages_top: LogMessageTopModel):
    logger.info(log_messages_top)
    await router_dest.broker.publish(
        log_messages_top.model_dump(by_alias=True),
        topic=settings['kafka_dest_topic']
    )


for topic in settings['kafka_souce_topics']:
    logger.info(f"Connected to {topic} as group {settings['kafka_source.username']}.{topic}.{settings['kafka_source.groupid']}")
    router_source.subscriber(
        topic,
        group_id=f"{settings['kafka_source.username']}.{topic}.{settings['kafka_source.groupid']}",
        auto_offset_reset="earliest"
    )(handle)

router_dest.include_router(router_source)
app = FastAPI(lifespan=router_dest.lifespan_context)
app.settings = settings
app.include_router(router_dest)

@router_dest.after_startup
async def on_startup(app: FastAPI):
    logger.info(f"Starting app on {Path().resolve().joinpath('app')}")
    await router_dest.broker.publish("Hi!", topic=settings['kafka_dest_topic'])
```

**Reply by Lancetnik** [reply]
@dimbler you are trying to get 2 different connections - it is the same case with multiple brokers. For now, **FastStream** doesn't support multiple brokers in the same time, but with FastAPI you can make it working.

Just create your custom lifespan to run both routers:

```python
from contextlib import asynccontextmanager

@asynccontextmanager
async def run_both(app):
    async with (
        router_dest.lifespan_context(app),
        router_source.lifespan_context(app),
    ):
        yield
 
app = FastAPI(lifespan=run_both)
```

**Reply by dimbler** [reply]
@Lancetnik this is really brilliant decision. Thank you very much for assistance.
Probably this decision have to be in user guide.

**Reply by Lancetnik** [reply]
@dimbler I don't like this solution so much. It doesn't merge both routers AsyncAPI schema, so far it's just a trick.

**Reply by danielriegert** [reply]
@Lancetnik is this still the recommended / only option to use faststream to connect to multiple brokers?
I need to connect to two different vhosts on the same RabbitMQ instance. However, it seems that the vhost can only be specified at the RabbitBroker level so I would have to create two RabbitBroker instances and use FastAPI as a wrapper?

**Reply by Lancetnik** [reply]
@danielriegert unfortunately, yes
For now FastStream doesn't support multiple brokers (we are working on it) and RabbitBroker uses the only one connection to RMQ, so u can specify vhost only at broker level

**Comment by danielriegert** [comment]
@Lancetnik Thanks for clarifying. Is there already a tentative date / timeline for supporting multiple brokers?

**Reply by Lancetnik** [reply]
I already fixed the main problem in logging in 0.5.0, but AsyncAPI documentation generation is still a job to be done

You can track this issue: https://github.com/airtai/faststream/issues/526

This is not a prior task, so I thought to impelement it in the end of the year, but we can review our priorities if requested
